raft/raft_test.go

```diff
@@ -4141,6 +4141,170 @@ func TestPreVoteMigrationWithFreeStuckPreCandidate(t *testing.T) {
+				{Term: 3, Index: 8},
+				{Term: 3, Index: 9},
+				{Term: 3, Index: 10},
+				{Term: 3, Index: 11},
+			},
+			2,
+			3,
+		},
+		// This case tests that leader can find the conflict index quickly.
+		// Firstly leader appends (type=MsgApp,index=8,logTerm=5, entries=...);
+		// After rejected leader appends (type=MsgApp,index=4,logTerm=3).
+		{
+			[]pb.Entry{
+				{Term: 1, Index: 1},
+				{Term: 2, Index: 2},
+				{Term: 2, Index: 3},
+				{Term: 3, Index: 4},
+				{Term: 4, Index: 5},
+				{Term: 4, Index: 6},
+				{Term: 4, Index: 7},
+				{Term: 5, Index: 8},
+			},
+			[]pb.Entry{
+				{Term: 1, Index: 1},
+				{Term: 2, Index: 2},
+				{Term: 2, Index: 3},
+				{Term: 3, Index: 4},
+				{Term: 3, Index: 5},
+				{Term: 3, Index: 6},
+				{Term: 3, Index: 7},
+				{Term: 3, Index: 8},
+				{Term: 3, Index: 9},
+				{Term: 3, Index: 10},
+				{Term: 3, Index: 11},
+			},
+			3,
+			4,
+		},
+		// This case tests that follower can find the conflict index quickly.
+		// Firstly leader appends (type=MsgApp,index=5,logTerm=1, entries=...);
+		// After rejected leader appends (type=MsgApp,index=2,logTerm=1).
+		{
+			[]pb.Entry{
+				{Term: 1, Index: 1},
+				{Term: 1, Index: 2},
+				{Term: 1, Index: 3},
+				{Term: 1, Index: 4},
+			},
+			[]pb.Entry{
+				{Term: 1, Index: 1},
+				{Term: 2, Index: 2},
+				{Term: 2, Index: 3},
+				{Term: 4, Index: 4},
+			},
+			1,
+			1,
+		},
+		// An normal case that there are no log conflicts.
+		// Firstly leader appends (type=MsgApp,index=5,logTerm=5, entries=...);
+		// After rejected leader appends (type=MsgApp,index=4,logTerm=4).
+		{
+			[]pb.Entry{
+				{Term: 1, Index: 1},
+				{Term: 1, Index: 2},
+				{Term: 1, Index: 3},
+				{Term: 4, Index: 4},
+				{Term: 5, Index: 5},
+			},
+			[]pb.Entry{
+				{Term: 1, Index: 1},
+				{Term: 1, Index: 2},
+				{Term: 1, Index: 3},
+				{Term: 4, Index: 4},
+			},
+			4,
+			4,
+		},
+	}
+
+	for i, test := range tests {
+		s1 := NewMemoryStorage()
+		s1.Append(test.ents1)
+		s2 := NewMemoryStorage()
+		s2.Append(test.ents2)
+
+		n1 := newTestRaft(1, []uint64{1, 2, 3}, 10, 1, s1)
+		n2 := newTestRaft(2, []uint64{1, 2, 3}, 10, 1, s2)
+
+		n1.becomeCandidate()
+		n1.becomeLeader()
+
+		n2.Step(pb.Message{From: 1, To: 1, Type: pb.MsgHeartbeat})
+
+		msgs := n2.readMessages()
+		if len(msgs) != 1 {
+			t.Errorf("can't read 1 message from peer 2")
+		}
+		if msgs[0].Type != pb.MsgHeartbeatResp {
+			t.Errorf("can't read heartbeat response from peer 2")
+		}
+		if n1.Step(msgs[0]) != nil {
+			t.Errorf("peer 1 step heartbeat response fail")
+		}
+
+		msgs = n1.readMessages()
+		if len(msgs) != 1 {
+			t.Errorf("can't read 1 message from peer 1")
+		}
+		if msgs[0].Type != pb.MsgApp {
+			t.Errorf("can't read append from peer 1")
+		}
+
+		if n2.Step(msgs[0]) != nil {
+			t.Errorf("peer 2 step append fail")
+		}
+		msgs = n2.readMessages()
+		if len(msgs) != 1 {
+			t.Errorf("can't read 1 message from peer 2")
+		}
+		if msgs[0].Type != pb.MsgAppResp {
```

Comment:
Do we want to assert on the contents of this `MsgAppResp`? At a minimum, we should check that its Reject field is true. But given the complex follower-side searching we do, it would also be interesting to test the `RejectHint` and `LogTerm` for each test case.

Reply:
Done.